### PR TITLE
[Breadth Coverage] Costmanagement

### PR DESCRIPTION
**Pull request by haroldrandom**
Close https://github.com/Azure/azure-cli/issues/12912

1. az costmanagement export create/update/list/show/delete
2. az costmanagement query

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Costmanagement

**Comment by jsntcy** [comment]
@haroldrandom, please fix CI errors first.

File: src/costmanagement/azext_costmanagement/manual/_params.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+def load_arguments(self, _):
+    with self.argument_context('costmanagement query') as c:
+        c.argument('scope',
+                   help='The scope associated with query and export operations. '
+                        'This includes "/subscriptions/{subscriptionId}/" for subscription scope, '
+                        '"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}" '
+                        'for resourceGroup scope, '
+                        '"/providers/Microsoft.Management/managementGroups/{managementGroupId}" '
+                        'for Management Group scope. ')
+        c.argument('type_',
```
Comment (MyronFanQiu) [inline src/costmanagement/azext_costmanagement/manual/_params.py:29]: >type_ [](start = 20, length = 5)

type_ ?

File: src/costmanagement/azext_costmanagement/manual/custom.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                                 definition_dataset_aggregation=None,
+                                 definition_dataset_grouping=None,
+                                 definition_dataset_filter=None,
+                                 schedule_status=None,
+                                 schedule_recurrence=None,
+                                 schedule_recurrence_period=None):
+
+    export_instance = client.get(scope=scope, export_name=export_name)
+
+    print(type(export_instance))
```
Comment (MyronFanQiu) [inline src/costmanagement/azext_costmanagement/manual/custom.py:96]: remove these two lines

File: src/costmanagement/azext_costmanagement/manual/custom.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        c.set_param('destination', delivery_info_destination)
+
+        # update export definition
+        c.set_param('timeframe', definition_timeframe)
+        c.set_param('configuration', definition_dataset_configuration)
+        c.set_param('aggregation', definition_dataset_aggregation)
+        c.set_param('grouping', definition_dataset_grouping)
+        c.set_param('filter', definition_dataset_filter)
+
+    print('-' * 100)
```
Comment (MyronFanQiu) [inline src/costmanagement/azext_costmanagement/manual/custom.py:126]: same here

File: src/costmanagement/azext_costmanagement/manual/custom.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                                   recurrence_period=export_instance.recurrence_period,
+                                   e_tag=export_instance.e_tag)
+
+
+def costmanagement_export_list(cmd, client, scope):
+    return client.list(scope=scope).value   # value exist even the result is empty
+
+
+def costmanagement_export_show(cmd, client,
+                               scope,
```
Comment (MyronFanQiu) [inline src/costmanagement/azext_costmanagement/manual/custom.py:150]: =None?

File: src/costmanagement/azext_costmanagement/manual/commands.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        operations_tmpl='azext_costmanagement.vendored_sdks.costmanagement.operations._export_operations#ExportOperatio'
+        'ns.{}',
+        client_factory=cf_export)
+    with self.command_group('costmanagement export', costmanagement_export, client_factory=cf_export,
+                            is_experimental=True) as g:
+        g.custom_command('list', 'costmanagement_export_list')
+        g.custom_show_command('show', 'costmanagement_export_show')
+        g.custom_command('create', 'costmanagement_export_create')
+        g.custom_command('update', 'costmanagement_export_update')
+        g.custom_command('delete', 'costmanagement_export_delete')
```
Comment (MyronFanQiu) [inline src/costmanagement/azext_costmanagement/manual/commands.py:31]: confirmation=True

Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/commands.py:31]: added.

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 4 above)
Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/custom.py:150]: This is copied from `generated/` without modification.

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 3 above)
Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/custom.py:126]: removed.

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 2 above)
Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/custom.py:96]: removed.

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 3 above)
Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/custom.py]: fixed

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 2 above)
Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/custom.py]: removed.

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 4 above)
Comment (MyronFanQiu) [reply src/costmanagement/azext_costmanagement/manual/custom.py:128]: The logic is a little bit strange for me. We checked the `scope` is equal to `None` or not. Should we modify it?

---
In reply to: [432214219](https://github.com/Azure/azure-cli-extensions/pull/1729#discussion_r432214219) [](ancestors = 432214219)

File: src/costmanagement/report.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+|Option|Type|Description|Path (SDK)|Path (swagger)|
+|------|----|-----------|----------|--------------|
+|**--scope**|string|The scope associated with alerts operations. This includes '/subscriptions/{subscriptionId}/' for subscription scope, '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}' for resourceGroup scope, '/providers/Microsoft.Billing/billingAccounts/{billingAccountId}' for Billing Account scope and '/providers/Microsoft.Billing/billingAccounts/{billingAccountId}/departments/{departmentId}' for Department scope, '/providers/Microsoft.Billing/billingAccounts/{billingAccountId}/enrollmentAccounts/{enrollmentAccountId}' for EnrollmentAccount scope, '/providers/Microsoft.Management/managementGroups/{managementGroupId} for Management Group scope, '/providers/Microsoft.Billing/billingAccounts/{billingAccountId}/billingProfiles/{billingProfileId}' for billingProfile scope, '/providers/Microsoft.Billing/billingAccounts/{billingAccountId}/billingProfiles/{billingProfileId}/invoiceSections/{invoiceSectionId}' for invoiceSection scope, and '/providers/Microsoft.Billing/billingAccounts/{billingAccountId}/customers/{customerId}' specific for partners.|scope|
```
Comment (MyronFanQiu) [inline src/costmanagement/report.md:9]: should we remove this file?

Comment (haroldrandom) [reply src/costmanagement/report.md:9]: removed.

File: src/costmanagement/azext_costmanagement/manual/custom.py (hunk 4 above)
Comment (haroldrandom) [reply src/costmanagement/azext_costmanagement/manual/custom.py:128]: Fixed

**Pull request merged**